gridfs.go

```diff
@@ -0,0 +1,601 @@
+package pail
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/gridfs"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// GridFSOptions support the use and creation of GridFS backed buckets.
+type GridFSOptions struct {
+	Name         string
+	Prefix       string
+	Database     string
+	MongoDBURI   string
+	DryRun       bool
+	DeleteOnSync bool
+	DeleteOnPush bool
+	DeleteOnPull bool
+	Verbose      bool
+}
+
+func (o *GridFSOptions) validate() error {
+	if (o.DeleteOnPush != o.DeleteOnPull) && o.DeleteOnSync {
+		return errors.New("ambiguous delete on sync options set")
+	}
+
+	return nil
+}
+
+type gridfsBucket struct {
+	opts   GridFSOptions
+	client *mongo.Client
+}
+
+func (b *gridfsBucket) normalizeKey(key string) string { return b.Join(b.opts.Prefix, key) }
+
+func (b *gridfsBucket) denormalizeKey(key string) string {
+	if b.opts.Prefix != "" && len(key) > len(b.opts.Prefix)+1 {
+		key = key[len(b.opts.Prefix)+1:]
+	}
+	return key
+}
+
+// NewGridFSBucketWithClient returns a new bucket backed by GridFS with the
+// existing Mongo client and given options.
+func NewGridFSBucketWithClient(ctx context.Context, client *mongo.Client, opts GridFSOptions) (Bucket, error) {
+	if client == nil {
+		return NewGridFSBucket(ctx, opts)
+	}
+
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	return &gridfsBucket{opts: opts, client: client}, nil
+}
+
+// NewGridFSBucket returns a bucket backed by GridFS with the given options.
+func NewGridFSBucket(ctx context.Context, opts GridFSOptions) (Bucket, error) {
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	client, err := mongo.NewClient(options.Client().ApplyURI(opts.MongoDBURI))
+	if err != nil {
+		return nil, errors.Wrap(err, "constructing client")
+	}
+
+	connctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+
+	err = client.Connect(connctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "connecting")
+	}
+
+	return &gridfsBucket{opts: opts, client: client}, nil
+}
+
+func (b *gridfsBucket) Check(ctx context.Context) error {
+	if b.client == nil {
```

Comment:
This nil check seems redundant given that the only entrypoints to create a GridFS bucket are to use the constructors, which already ensure that a client is initialized before returning the bucket.